File: bridge/model/vmexplorer.go
package model

import (
	"bridge/utils"
	"fmt"
)

const VMENumEntries = 512

const VMExplorerUnspecIndex int32 = -1

type VMEPath struct {
	L4 int32 `json:"l4"`
	L3 int32 `json:"l3"`
	L2 int32 `json:"l2"`
	L1 int32 `json:"l1"`
}

type NetlinkVMEntry struct {
	Value    uint64
	PA       uint64
	KernelVA uint64
	UserVA   uint64
	Size     uint64
	Present  bool
	Bad      bool
	Leaf     bool
	None     bool
}

func (e NetlinkVMEntry) String() string {
	return fmt.Sprintf("0x%016x [PA=0x%016x, KernelVA=0x%016x, UserVA=0x%016x]", e.Value, e.PA, e.KernelVA, e.UserVA)
}

func ReadNetlinkVMEntry(parser utils.ByteParser) (NetlinkVMEntry, error) {
	var entry NetlinkVMEntry

	entry.Value = parser.ReadUint64()
	entry.PA = parser.ReadUint64()
	entry.KernelVA = parser.ReadUint64()
	entry.UserVA = parser.ReadUint64()
	entry.Size = parser.ReadUint64()
	entry.Present = parser.ReadBool()
	entry.Bad = parser.ReadBool()
	entry.Leaf = parser.ReadBool()
	entry.None = parser.ReadBool()

	return entry, parser.Error()
}

type NetlinkVMEDump struct {
	Entries []NetlinkVMEntry
}

type WebsocketVMEntry struct {
	Index       int32  `json:"index"`
	RawValue    string `json:"rawValue"`
	PA          string `json:"pa"`
	KernelVA    string `json:"kernelVA"`
	UserVAStart string `json:"userVAStart"`
	UserVAEnd   string `json:"userVAEnd"`
	Size        string `json:"size"`
	Leaf        bool   `json:"leaf"`
	Present     bool   `json:"present"`
	None        bool   `json:"none"`
}

func ToWebsocketVMEntry(entry NetlinkVMEntry, index int) WebsocketVMEntry {
	return WebsocketVMEntry{
		Index:       int32(index),
		RawValue:    fmt.Sprintf("0x%016x", entry.Value),
		PA:          fmt.Sprintf("0x%016x", entry.PA),
		KernelVA:    fmt.Sprintf("0x%016x", entry.KernelVA),
		UserVAStart: fmt.Sprintf("0x%016x", entry.UserVA),
		UserVAEnd:   fmt.Sprintf("0x%016x", entry.UserVA+entry.Size),
		Size:        fmt.Sprintf("0x%016x", entry.Size),
		Leaf:        entry.Leaf,
		Present:     entry.Present,
		None:        entry.None,
	}
}

type WebsocketVMEReq struct {
	Key  TaskKey `json:"key"`
	Path VMEPath `json:"path"`
}

type WebsocketVMEDump struct {
	Entries []WebsocketVMEntry `json:"entries"`
}

File: webapp/src/types/ws/vmexplorer.ts
import type { WebsocketTaskKey } from "./shared";

export const VME_UNSPEC_INDEX: number = -1;

export type WebsocketVMEntry = {
  index: number;
  rawValue: string;
  pa: string;
  kernelVA: string;
  userVAStart: string;
  userVAEnd: string;
  size: string;
  leaf: boolean;
  present: boolean;
  bad: boolean;
  none: boolean;
};

export function isWebsocketVMEntry(obj: any): obj is WebsocketVMEntry {
  return (
    obj !== null &&
    typeof obj === "object" &&
    typeof obj.rawValue === "string" &&
    typeof obj.pa === "string" &&
    typeof obj.kernelVA === "string" &&
    typeof obj.userVAStart === "string" &&
    typeof obj.userVAEnd === "string" &&
    typeof obj.size === "string" &&
    typeof obj.leaf === "boolean" &&
    typeof obj.present === "boolean" &&
    typeof obj.bad === "boolean" &&
    typeof obj.none === "boolean"
  );
}

export type WebsocketVMEDump = {
  entries: WebsocketVMEntry[];
};

export function isWebsocketVMEDump(obj: any): obj is WebsocketVMEDump {
  return (
    obj !== null &&
    typeof obj === "object" &&
    Array.isArray(obj.entries) &&
    obj.entries.every(isWebsocketVMEntry)
  );
}

export type VMEPath = {
  l4: number;
  l3: number;
  l2: number;
  l1: number;
};

export type WebsocketVMEReq = {
  key: WebsocketTaskKey;
  path: VMEPath;
};
